py3status/py3.py

```diff
@@ -1008,3 +1009,165 @@ def request(self, url, params=None, data=None, headers=None,
+
+        :param history: history of the variable
+        :param value: value to add to the history
+        :param length: length of the history
+        :param default_value: default value to pad with
+        :param minimum: minimum value possible
+        :param maximum: maximum value possible
+        :param bars: characters to use for the different levels of the bar
+        :param separator: character to insert between the bars
+
+        :returns: Composite list
+        """
+
+        if bars is None:
+            bars = self._default_bars
+
+        while len(history) < length:
+            history.append(default)
+
+        history.append(value)
+
+        while len(history) > length:
+            history = history[1:]
+
+        return self.composite_join(separator,
+                                   [self._value_to_bar(x,
+                                                       minimum,
+                                                       maximum,
+                                                       bars)
+                                    for x in history])
+
+    def progress_bar(self, value, length=8, middle_char='|', left_char='─',
+                     right_char='─', middle_color=None, left_color=None,
+                     right_color=None, minimum=0, maximum=100, separator=''):
+        """
+        Transform a value into a progress bar
+
+        :param value: value to convert to a progress bar
+        :param length: length of the progress bar
+        :param middle_char: character to use for the middle marker
+        :param left_char: character to use for the characters on the left of
+                          the marker
+        :param right_char: character to use for the characters on the right of
+                           the marker
+        :param middle_color: color to use for the middle marker, if set to
+                             None, will use thresholds instead
+        :param left_color: color to use for the characters on the left of the
+                           marker, if set to None, will use thresholds instead
+        :param right_color: color to use for the character on the right of the
+                            marker, if set to None, will use thresholds instead
+        :param minimum: minimum value possible
+        :param maximum: maximum value possible
+        :param separator: character to insert between the bars
+
+        :returns: Composite list
+        """
+
+        progress = []
+
+        idx = self._value_to_index(value, length, minimum, maximum)
+
+        for i in range(idx - 1):
+            i_val = self._index_to_value(i, length, minimum, maximum)
+            if left_color is None:
+                color = self.threshold_get_color(i_val)
+            else:
+                color = left_color
+            progress.append({'full_text': left_char, 'color': color})
+
+        if middle_color is None:
+            color = self.threshold_get_color(value)
+        else:
```

Comment:
We waste 20 loops doing `i_val = self._index_to_value(i, l, m, m)` if `left_color` is not `None`.
We waste 20 loops doing `i_val = self._index_to_value(i, l, m, m)` if `right_color` is not `None`.

Let's check the colors first before we do anything with `i_val`.